src/Npgsql/ConnectorPool.cs

```diff
@@ -65,39 +87,46 @@ internal ConnectorPool(NpgsqlConnectionStringBuilder settings, string connString
+                throw new ArgumentException("ConnectionPruningInterval can't be 0.");
+            var connectionIdleLifetime = TimeSpan.FromSeconds(settings.ConnectionIdleLifetime);
+            var pruningSamplingInterval = TimeSpan.FromSeconds(settings.ConnectionPruningInterval);
+            if (connectionIdleLifetime < pruningSamplingInterval)
+                throw new ArgumentException($"Connection can't have ConnectionIdleLifetime {connectionIdleLifetime} under ConnectionPruningInterval {_pruningSamplingInterval}");
+
+            _pruningTimer = new Timer(PruningTimerCallback, this, Timeout.Infinite, Timeout.Infinite);
+            _pruningSampleSize = DivideRoundingUp(settings.ConnectionIdleLifetime, settings.ConnectionPruningInterval);
+            _pruningMedianIndex = DivideRoundingUp(_pruningSampleSize, 2) - 1; // - 1 to go from length to index
+            _pruningSamplingInterval = pruningSamplingInterval;
+            _pruningSamples = new int[_pruningSampleSize];
+            _pruningTimerEnabled = false;
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal bool TryAllocateFast(NpgsqlConnection conn, [NotNullWhen(true)] out NpgsqlConnector? connector)
+        internal ValueTask<NpgsqlConnector> Rent(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async,
+            CancellationToken cancellationToken)
         {
-            if (TryGetIdleConnector(out connector))
+            if (TryGetIdleConnector(out var connector))
             {
                 connector.Connection = conn;
-                return true;
+                return new ValueTask<NpgsqlConnector>(connector);
             }
 
-            return false;
-        }
+            return RentAsync(conn, timeout, async, cancellationToken);
 
-        internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async,
-            CancellationToken cancellationToken)
-        {
-            NpgsqlConnector? connector = null;
-
-            // No idle connector was found in the pool.
-            // We now loop until one of three things happen:
-            // 1. The pool isn't at max capacity (Open < Max), so we can create a new physical connection.
-            // 2. The pool is at maximum capacity and there are no idle connectors (Open - Idle == Max),
-            // so we enqueue an open attempt into the waiting queue, so that the next release will unblock it.
-            // 3. An connector makes it into the idle list (race condition with another Release()).
-            while (true)
+            async ValueTask<NpgsqlConnector> RentAsync(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async,
+                CancellationToken cancellationToken)
             {
-                var numConnectors = _numConnectors;
-                if (numConnectors < _max)
+                // TODO: If we're synchronous, use SingleThreadSynchronizationContext to not schedule completions
```

Comment:
Note: the comment is stale - SingleThreadSynchronizationContext is already implemented below (when using ReadAsync, when the pool is exhausted).

> If the underlying operation is async you'd always still need a thread to wake up the task and post the continuation to the sync context (this could be say an IOCP, epoll or threadpool thread)

According to my current mental model, the crucial part is whether a TP thread is necessary at any point to complete the operation - here's the problematic scenario. If a lot of connections are suddenly being synchronously opened by TP threads (e.g. thundering herd), the TP gets exhausted and slowly starts ramping up. At the same time, if a TP thread is needed to complete these operations, none are available and we're in a pseudo-deadlock. All the SingleThreadSynchronizationContext does, is to move the completion to some arbitrary non-TP thread to prevent that.

PS You may want to give https://github.com/npgsql/npgsql/pull/2975#issuecomment-633055676 a read on another similar usage of SingleThreadSynchronizationContext. If sync-over-async can somehow complete without requiring any TP thread, I'd be very happy to know that and it would be great to get rid of this stuff.

Reply:
Another thing this mitigates, if I'm not mistaken, is if a UI program performs a sync operation. If we didn't have SingleThreadSynchronizationContext, wouldn't the completion be scheduled to execute on the UI thread, which is synchronously blocked?

Reply:
>According to my current mental model, the crucial part is whether a TP thread is necessary at any point to complete the operation

This depends what kind the completing thread is, generally these IO threads don't want user code running on them so you can expect RunContinuationsAsync (or similar) at the least. Continuations that supplied a sync context, under normal circumstances, would be able to skip the TP posting straight from the IO thread into the SyncCtx. For epoll I'm not 100% as it's weirder in its resumption semantics; it apparently always resumes on a TP thread, meaning you'd go from TP to SyncCtx and 'back' to a sync blocked TP thread.

>completion be scheduled to execute on the UI thread

Sure, you need this or `ConfigureAwait(false)` sprinkled all over npgsql internals :)

Reply:
Did you get a definite answer on the epoll threads?

Reply:
Nope. But I guess the easiest thing is to code up a quick test, i.e. throw lots of tasks to the TP which block synchronously, once with a special sync context (like the one Npgsql uses) and one without. Let me know if you feel like doing that, or I can do it. Also just as a reminder, we still support netfx and also older versions of netcore, which could also be a factor.